### ISSUE TITLE
Explore new loss functions

### ISSUE THREAD

**Issue by ebgoldstein** [Doodleverse/segmentation_gym#74]
For multiclass problems where there is class imbalance, Dice often works well as the loss function. For some problems this is not perfect, and a rare class might not be predicted well enough. Looking around in the literature (specifically medical image segmentation), I found some suggestions that a different loss function might be needed.. even a compound loss. 

In new versions of Gym it would be nice to have a loss formulation that really focuses on rare classes and poorly segmented examples. Based on this [repository](https://github.com/JunMa11/SegLoss) and [paper](https://doi.org/10.1016/j.media.2021.102035) - 

i think there might be  new losses:

-Focal Loss
-topK loss
-dice loss + Focal
-dice loss + topK

example of focal for segmentation in TF can be seen [here in this repository](https://github.com/maxvfischer/keras-image-segmentation-loss-functions/blob/master/losses/multiclass_losses.py)

I know the loss formulations are in `doodleverse_utils`, and will be added there.. but this is a gym issue, so i will post it here

**Comment by MARDAScience** [comment]
Agree and I also think a systematic test against a benchmark is useful for testing existing functions such as Hinge, as well as new loss functions. The benchmark data should ideally include

1. well-balanced binary data
2. imbalanced binary data
3. well-balanced multiclass data
4. imbalanced multiclass data

Finally, what is a metric to measure class imabalance? It would be nice to plot "amount of imbalance" against "performance" for all loss functions used on the same data

**Comment by ebgoldstein** [comment]
see here: https://github.com/Doodleverse/doodleverse_utils/issues/7#issuecomment-1263021118

maybe loss weighting is needed more than a zillion loss function options...

**Comment by dbuscombe-usgs** [comment]
Loss weighting could be passed as a vector of length=NCLASSES. Could be ad hoc vector of numbers of proportions, or generated from the train/validation datasets.

**Comment by ebgoldstein** [comment]
Loss weighting is now implemented.. Should we close this issue for now as 'not planned', or keep it open and eventually explore more loss functions for the various datasets listed above? i.e., for a short technical note akin to the 'Loss odyssey' paper?

**Comment by dbuscombe-usgs** [comment]
keep this open as a space to discuss what we see when we use loss weighting? Not been thoroughly tested/benchmarked yet

**Comment by dbuscombe-usgs** [comment]
Closing this issue. Making a new issue 'explore loss weighting'

**Comment by ebgoldstein** [comment]
I wanted to close this as not planned at the moment... I still am interested in Focal loss, TopK, and these combo loss functions... just not now that Dice is working so well...
